itwm_example/mco/mco.py

```diff
@@ -118,109 +77,114 @@ def optimize(self):
+        parameters = model.parameters
+        kpis = model.kpis
 
+        if model.evaluation_mode == "Subprocess":
+            # Here we create an instance of our WorkflowEvaluator subclass
+            # that allows for evaluation of a state in the workflow via calling
+            # force_bdss on a new subprocess running in 'evaluate' mode.
+            # Note: a BaseMCOCommunicator must be present to pass in parameter
+            # values and returning the KPI for a force_bdss run in 'evaluate'
+            # mode
+            single_point_evaluator = SubprocessWorkflowEvaluator(
+                workflow=evaluator.workflow,
+                workflow_filepath=evaluator.workflow_filepath,
+                executable_path=sys.argv[0],
+            )
+        else:
+            single_point_evaluator = evaluator
 
-def _int_weights(dimension, num_points, zero_values):
-    """Helper routine for the previous one. The meaning is the same, but
-    works with integers instead of floats, adding up to num_points"""
+        #: Get scaling factors and non-zero weight combinations for each KPI
+        scaling_factors = self.get_scaling_factors(
+            single_point_evaluator, kpis, parameters
+        )
 
-    if dimension == 1:
-        yield [num_points - 1]
-    else:
-        if zero_values:
-            integers = np.arange(num_points-1, -1, -1)
-        else:
-            integers = np.arange(num_points-2, 0, -1)
-        for i in integers:
-            for entry in _int_weights(dimension - 1,
-                                      num_points - i,
-                                      zero_values):
-                yield [i] + entry
-
-
-def get_scaling_factors(single_point_evaluator, kpis, parameters):
-    """KPI Scaling factors for MCO are calculated (as required) by
-    normalising by the possible range of each optimal KPI value.
-    Also known as Sen's Multi-Objective Programming Method[1]_.
-
-    References
-    ----------
-    .. [1] Chandra Sen, "Sen's Multi-Objective Programming Method and Its
-       Comparison with Other Techniques", American Journal of Operational
-       Research, vol. 8, pp. 10-13, 2018
-    """
+        for weights in model.weights_samples(with_zero_values=False):
 
-    #: Get initial weights referring to extrema of each variable range
-    auto_scales = [kpi.auto_scale for kpi in kpis]
-    scaling_factors = [kpi.scale_factor for kpi in kpis]
-    extrema = np.zeros((len(kpis), len(kpis)))
-    initial_weights = get_weight_combinations(len(kpis), 2)
+            log.info("Doing MCO run with weights: {}".format(weights))
 
-    #: Calculate extrema for each KPI optimisation
-    for i, weights in enumerate(initial_weights):
+            generator = zip(weights, scaling_factors)
+            scaled_weights = [weight * scale for weight, scale in generator]
 
-        log.info("Doing extrema MCO run with weights: {}".format(weights))
+            evaluator = self.optimizer(
```

Comment:
Do we need to create multiple instances of `self.optimizer` here, or could we instead create one instance outside the for loop and then just update the weights during each iteration?